New slur adjustment algorithm

DavidBauer1984 opened the pull request:
This PR implements a new slur adjustment algorithm that mainly adjusts the control points of the slur bezier curve.

Slurs now more accurately curve around obstacles while preserving the position of the slur endpoints as much as possible.

| Before | After |
| ------ | ----- |
| ![Before1](https://user-images.githubusercontent.com/63608463/133742968-40177d29-d58f-48c3-a6eb-abf6b3517573.png) | <img width="219" alt="After1" src="https://user-images.githubusercontent.com/63608463/133743026-9094a5d9-5e5f-4bd1-b356-34163f1da143.png"> |

**Before**
<img width="1533" alt="Before2" src="https://user-images.githubusercontent.com/63608463/133743711-f6f18bd7-4e85-4d32-a0d7-0bfb7c6ff6c6.png">
**After**
<img width="1546" alt="After2" src="https://user-images.githubusercontent.com/63608463/133743751-577f342e-df35-43ce-9339-9d55ad893df4.png">

This particularly improves long slurs where endpoints before appeared lifted off.

**Before**
<img width="800" alt="Before3" src="https://user-images.githubusercontent.com/63608463/133744015-7eb5aaba-22f3-4d8e-a804-3c49060551b6.png">
<img width="500" alt="Before4" src="https://user-images.githubusercontent.com/63608463/133744052-af1d446f-ab4c-49f6-9f5c-ef8d7d114300.png">

**After**
<img width="800" alt="After3" src="https://user-images.githubusercontent.com/63608463/133744089-fabb60a1-2acc-464a-9ae3-4fa7d52ddc8f.png">
<img width="500" alt="After4" src="https://user-images.githubusercontent.com/63608463/133744124-9c59131d-91f4-4e0f-b286-5a6a812556dd.png">

The new algorithm consists of three steps:

1. Vertical end point adjustment. This shifts the slur vertically. Only collisions near the endpoints are taken into account.
2. Horizontal control point adjustment. The idea is to shift control points to the outside if there is an obstacle in the vicinity of the corresponding endpoint.
3. Vertical control point adjustment. For each colliding bounding box we formulate a constraint `ax + by >= c` where x, y denote the vertical adjustments of the control points and c is the size of the collision. The coefficients a, b are calculated from the Bezier curve equation. After collecting all constraints we calculate a solution.

<img width="600" alt="AlgoSteps" src="https://user-images.githubusercontent.com/63608463/133747895-f009ee9b-d6df-4bf6-a2d7-c3caded86f3d.png">

Moreover:
- We fix an error in the calculation of `t` in `BoundingBox::CalcBezierAtPosition`. This error means that previously the slur was evaluated at the wrong horizontal position in order to calculate intersections with bounding boxes. To accurately calculate `t` one has to solve a cubic equation. For now we determine an approximate solution by interval bisection, this might be refined in a later PR.
- We adjust cross staff slurs after vertical justification. This fixes part of #2309 .

lpugin commented:
This looks excellent. Also many thanks for the thorough explanations. It is nice to have the slurs remaining closer to the content and not remaining symmetrical. That gives them a more natural shape. 

A general remark on the appearance. In some cases, the change make them look not enough curved at the end and a bit too pointy in my opinion. This happens also in cases where there is no obstacle that should modify the shape. Maybe a bit more spaces between the control points would help this? See slur-009
![image](https://user-images.githubusercontent.com/689412/134289608-2a302528-8a65-4113-b3c9-2da5c2ca3f41.png)

Some comments for specific cases

**slur-005** - The look good, but any ideas why it is cutting one stem?
![image](https://user-images.githubusercontent.com/689412/134290321-f9f717f4-e1df-4675-a31c-5d35e9239ffc.png)

**dynam-008** - Is the change in the dynamic placement related? (same in **metersig-002**)
![image](https://user-images.githubusercontent.com/689412/134290074-95e2e283-a5b0-4291-92a0-e18f8c0bf992.png)

**gracenote-017** - The second slur in the lower staff is now touching the note head
![image](https://user-images.githubusercontent.com/689412/134290147-28a1b5cc-a675-4301-a043-dbb64b81ab2d.png)

DavidBauer1984 commented:
Thank you for the examples. I'll discuss them one by one.

**Slur-009**
<img width="409" alt="Slur1" src="https://user-images.githubusercontent.com/63608463/134638072-24705aa5-b0d3-4f2b-ae61-dc328d8b3466.png">

No big change here, but if you want slurs to appear more curvy, you can increase the slur curve factor. Same example rendered with `--slur-curve-factor 25`:

<img width="459" alt="Slur2" src="https://user-images.githubusercontent.com/63608463/134638509-cb21cb6a-b44e-4760-9b15-c588007c7a6d.png">

Also the end point of this slur could be higher, but I do not want to change this here, since this PR is about slur adjustment. The true slur end point positions are now more visible (since endpoints are less often adjusted). This also means that some suboptimal end point positions come to light which should be addressed in a follow-up-PR.

**CrossStaff-005**
Fixed.
<img width="1098" alt="Slur3" src="https://user-images.githubusercontent.com/63608463/134639342-8c6f802b-1b99-4df9-8854-a771a4973823.png">

**Dynam-008**
Yeah this was caused by the change in `FloatingPositioner::SetYRel(...)`. Fixed as well.
<img width="888" alt="Dynam" src="https://user-images.githubusercontent.com/63608463/134639633-dfc01d28-d4f0-472c-aff3-6c8d6943fd41.png">

**Gracenote-017**
No change here.
<img width="552" alt="Slur4" src="https://user-images.githubusercontent.com/63608463/134639871-711121b1-2e57-4ff8-8105-c1c44ce31376.png">
This is again concerning slur endpoints and might already look better once #2367 is merged.

craigsapp commented:
Here is a survey of slur layout from various hand-engraved version of this example:

**Before**
<img width="1533" alt="Before2" src="https://user-images.githubusercontent.com/63608463/133743711-f6f18bd7-4e85-4d32-a0d7-0bfb7c6ff6c6.png">
**After**
<img width="1546" alt="After2" src="https://user-images.githubusercontent.com/63608463/133743751-577f342e-df35-43ce-9339-9d55ad893df4.png">

Notice that the slurs in mm. 103–104 in the hand engravings are closer in shape and position to the **Before** version above:

<details>
<summary>
Breitkof und Härtel 1878
</summary>

[Friedrich Chopin's Werke](https://imslp.org/wiki/Friedrich_Chopin%27s_Werke_(Chopin,_Fr%C3%A9d%C3%A9ric)). Band IX (pp.34-45)
Leipzig: [Breitkopf und Härtel](https://imslp.org/wiki/Breitkopf_und_H%C3%A4rtel), n.d.[1878]. Plate C. IX. 6–8.

https://s9.imslp.org/files/imglnks/usimg/d/d7/IMSLP114892-PMLP02373-FChopin_Waltzes,_Op.64_BH9.pdf

</details>

<img width="751" alt="Screen Shot 2022-04-08 at 10 36 57 PM" src="https://user-images.githubusercontent.com/3487289/162558067-fc74d7e2-c78e-4c1e-be32-29b599e3b944.png">


<details>
<summary>
Schirmer 1880, Kullak edition
</summary>

Klavierwerke. Instructive Ausgabe, Vol.VI: WaltzesBerlin: Schlesinger'sche Buch-und Musikhandlung, 1880. Plate S. 7291(6-8).Co-issue - New York: G. Schirmer, 1880.

https://s9.imslp.org/files/imglnks/usimg/e/ef/IMSLP82854-PMLP02373-FChopin_Waltzes_Op64.pdf

</details>

<img width="822" alt="Screen Shot 2022-04-08 at 10 41 48 PM" src="https://user-images.githubusercontent.com/3487289/162558226-69b73ca4-1bf6-43cf-bd29-3843641868b3.png">


<details>
<summary>
Schirmer 1894, Mikuli edition
</summary>

[Complete Works for the Piano](https://imslp.org/wiki/Complete_Works_for_the_Piano,_1894_(Chopin,_Fr%C3%A9d%C3%A9ric)), Vol.1: Waltzes (pp.40-54)
New York: [G. Schirmer](https://imslp.org/wiki/Schirmer), 1894. Plate 11082-11084.

https://s9.imslp.org/files/imglnks/usimg/7/74/IMSLP00499-Chopin_-_Waltzes,_Op_64.pdf

</details>

<img width="734" alt="Screen Shot 2022-04-08 at 10 44 46 PM" src="https://user-images.githubusercontent.com/3487289/162558314-aebbd5c5-82a7-482e-bb11-c077d91373cf.png">

<details>
<summary>
Peters 1905, Scholtz edition
</summary>

Sämtliche Pianoforte-Werke, Band 1 (pp.30-39)Leipzig: C.F. Peters, n.d.[(ca.1905)]. Plate 9462.

https://s9.imslp.org/files/imglnks/usimg/6/64/IMSLP51544-PMLP02373-Chopin_Klavierwerke_Band_1_Peters_9462_Op.64_600dpi_scan.pdf

</details>

<img width="757" alt="Screen Shot 2022-04-08 at 10 47 41 PM" src="https://user-images.githubusercontent.com/3487289/162558399-9ebcf3b3-b81f-4e1d-909f-53d6c9980af2.png">


<details>
<summary>
Durand 1915, Debussy edition
</summary>
Œuvres completes pour piano: Valses (pp.38-53)
Paris: [Durand & Fils](https://imslp.org/wiki/Durand), 1915. Plate D. & F. 9709.

https://s9.imslp.org/files/imglnks/usimg/8/8b/IMSLP80377-PMLP02373-Chopin_Valses_Durand_9709_Op_64_filter.pdf

</details>

<img width="756" alt="Screen Shot 2022-04-08 at 10 49 20 PM" src="https://user-images.githubusercontent.com/3487289/162558463-95c38ead-1f9a-40ed-a2df-655c22e62ede.png">

<img width="751" alt="Screen Shot 2022-04-08 at 10 50 11 PM" src="https://user-images.githubusercontent.com/3487289/162558495-91d23c08-8899-4966-866b-36a3391393bc.png">

<details>
<summary>
Schirmer 1915, Joseffy edition
</summary>
[Complete Works for the Piano](https://imslp.org/wiki/Complete_Works_for_the_Piano,_1915_(Chopin,_Fr%C3%A9d%C3%A9ric)), Vol.1: Waltzes (LMC 27) (pp.40-54)
New York: [G. Schirmer](https://imslp.org/wiki/Schirmer), 1915. Plate 25502..

https://s9.imslp.org/files/imglnks/usimg/b/b6/IMSLP167283-PMLP02373-FChopin_Waltzes,_Op.64_Joseffy.pdf

</details>

<img width="784" alt="Screen Shot 2022-04-08 at 10 54 19 PM" src="https://user-images.githubusercontent.com/3487289/162558632-04931439-31a0-4aec-bfba-bc6780481f54.png">


<details>
<summary>
Polskie Wydawnictwo Muzyczne, 1949  Paderewski edition: This one is more similar to the <b>After</b> sample.
</summary>

	Dzieła wszystkie Fryderyka Chopina, Vol.IX: Waltzes (pp.55-70)
Warszawa: Polskie Wydawnictwo Muzyczne, 1949. Plate PWM 239. (1956 reissue)

https://imslp.eu/files/imglnks/euimg/4/4c/IMSLP102214-PMLP02373-Chopin_Werke_Paderewski_09_Walzer_239_06-08_Op_64_scan.pdf

</details>

<img width="777" alt="Screen Shot 2022-04-08 at 10 56 47 PM" src="https://user-images.githubusercontent.com/3487289/162558711-a8e802dc-4bc3-4eb7-967a-e73199b72c2e.png">









Examples from [IMSLP](https://imslp.org/wiki/Waltzes,_Op.64_(Chopin,_Fr%C3%A9d%C3%A9ric)).
